wallet-new/src/Cardano/Wallet/Kernel/CoinSelection/FromGeneric.hs

```diff
@@ -377,27 +388,52 @@ largestFirst opts maxInps =
            second 1 is the encoded address type. 30 = size of Blake2b_224
            hash of Address'.
 -}
-estimateSize :: Int      -- ^ Average size of @Attributes AddrAttributes@.
-             -> Int      -- ^ Size of transaction's @Attributes ()@.
+estimateSize :: Byte     -- ^ Average size of @Attributes AddrAttributes@.
+             -> Byte     -- ^ Size of transaction's @Attributes ()@.
              -> Int      -- ^ Number of inputs to the transaction.
              -> [Word64] -- ^ Coin value of each output to the transaction.
              -> Byte     -- ^ Estimated size of the resulting transaction.
-estimateSize saa sta ins outs
-    = fromBytes . fromIntegral $
-      5
-    + 42 * ins
-    + (11 + listSize (32 + (fromIntegral saa))) * length outs
-    + sum (map intSize outs)
-    + fromIntegral sta
+estimateSize saa sta ins outs =
+    case szSimplify (szWithCtx (Map.fromList ctx) (Proxy @TxAux)) of
+        Left  sz    -> error ("Size estimate failed to simplify: " <> pretty sz)
```

Comment:
Since this can give an error, I'd *really* prefer to see this whole function return `Either` as well. An `unsafeEstimateSize` that called `error` can be used for backwards compatibility.

Reply:
This is actually a little weird, and I'd love to figure out the best approach. It's true that this function could return `Left`, but only if the structure of a `TxAux` changes in such a way that the symbolic formula can't simplify to a value (say, because it suddenly starts depending on the encoded size of some type `T` that we haven't handled yet).

And there is a test in `core` that ensures the expression for `TxAux` really does reduce to a range value, and the range encloses the true encoded size: https://github.com/input-output-hk/cardano-sl/blob/develop/core/test/Test/Pos/Core/Bi.hs#L1809

So in one sense it doesn't make sense to bubble up an error, because the linked test is supposed to catch any change to `TxAux` that could cause a `Left` here. But on the other hand, this connection *does* feel slightly more tenuous than I'd have liked...

Reply:
Would you be willing to put this justification in the file as a comment? I'm satisfied that it's probably safe now.